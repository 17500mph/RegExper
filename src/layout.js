import nodeTypes from 'rendering/types';

const normalizeBBox = box => ({
  width: 0,
  height: 0,
  axisY: (box.height || 0) / 2,
  axisX1: 0,
  axisX2: (box.width || 0),
  ...box
});

const layout = data => {
  if (typeof data == 'string') {
    return data;
  }

  const { type } = data;

  if (data.children) {
    data.children = data.children.map(layout);
  }

  const normalizedData = {
    props: {},
    box: {},
    ...data
  };
  const result = nodeTypes[type].layout(normalizedData) || normalizedData;

  return {
    ...result,
    box: normalizeBBox(result.box)
  };
};

export default layout;
