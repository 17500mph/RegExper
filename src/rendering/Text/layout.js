import React from 'react';
import ReactDOM from 'react-dom';

import Text from 'rendering/Text';

const layoutText = data => {
  const container = document.createElement('div');
  document.body.appendChild(container);

  ReactDOM.render(
    <svg width="0" height="0" viewBox="0 0 0 0">
      <Text { ...data.props }>{ data.children }</Text>
    </svg>,
    container);

  const box = container.querySelector('svg > text').getBBox();
  document.body.removeChild(container);

  data.box = {
    width: box.width,
    height: box.height
  };
  data.props.transform = `translate(${ -box.x } ${ -box.y })`;
  return data;
};

export default layoutText;
