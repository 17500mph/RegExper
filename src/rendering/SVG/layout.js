import layout from 'layout';

const layoutSVG = data => {
  const child = layout(data.children[0]);

  data.props.innerWidth = child.box.width;
  data.props.innerHeight = child.box.height;

  return data;
};

export default layoutSVG;
