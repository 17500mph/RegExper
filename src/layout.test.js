import layout from 'layout';

jest.mock('rendering/types', () => ({
  Example: {
    layout: jest.fn()
  },
  Other: {
    layout: jest.fn()
  }
}));

import nodeTypes from 'rendering/types';

describe('layout', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('running against a string', () => {
    expect(layout('test')).toEqual('test');
  });

  test('running against a simple node', () => {
    expect(layout({
      type: 'Example'
    })).toMatchSnapshot();
  });

  test('running against a node with a basic bounding box', () => {
    nodeTypes.Example.layout.mockImplementation(data => {
      data.box = { width: 20, height: 10 };
      return data;
    });
    expect(layout({
      type: 'Example'
    })).toMatchSnapshot();
  });

  test('running against a node with a complete bounding box', () => {
    nodeTypes.Example.layout.mockImplementation(data => {
      data.box = {
        width: 20,
        height: 10,
        axisY: 2,
        axisX1: 5,
        axisX2: 15
      };
      return data;
    });
    expect(layout({
      type: 'Example'
    })).toMatchSnapshot();
  });

  test('running against a node with props', () => {
    expect(layout({
      type: 'Example',
      props: {
        property: 'example'
      }
    })).toMatchSnapshot();
  });

  test('running layout on children', () => {
    nodeTypes.Other.layout.mockImplementation(data => {
      data.box = { width: 20, height: 10 };
      return data;
    });
    expect(layout({
      type: 'Example',
      children: [
        {
          type: 'Other'
        },
        'string example'
      ]
    })).toMatchSnapshot();
  });
});
