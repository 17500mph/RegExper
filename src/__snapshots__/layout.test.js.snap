// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout running against a node with a basic bounding box 1`] = `
Object {
  "box": Object {
    "axisX1": 0,
    "axisX2": 20,
    "axisY": 5,
    "height": 10,
    "width": 20,
  },
  "props": Object {},
  "type": "Example",
}
`;

exports[`layout running against a node with a complete bounding box 1`] = `
Object {
  "box": Object {
    "axisX1": 5,
    "axisX2": 15,
    "axisY": 2,
    "height": 10,
    "width": 20,
  },
  "props": Object {},
  "type": "Example",
}
`;

exports[`layout running against a node with props 1`] = `
Object {
  "box": Object {
    "axisX1": 0,
    "axisX2": 0,
    "axisY": 0,
    "height": 0,
    "width": 0,
  },
  "props": Object {
    "property": "example",
  },
  "type": "Example",
}
`;

exports[`layout running against a simple node 1`] = `
Object {
  "box": Object {
    "axisX1": 0,
    "axisX2": 0,
    "axisY": 0,
    "height": 0,
    "width": 0,
  },
  "props": Object {},
  "type": "Example",
}
`;

exports[`layout running layout on children 1`] = `
Object {
  "box": Object {
    "axisX1": 0,
    "axisX2": 0,
    "axisY": 0,
    "height": 0,
    "width": 0,
  },
  "children": Array [
    Object {
      "box": Object {
        "axisX1": 0,
        "axisX2": 20,
        "axisY": 5,
        "height": 10,
        "width": 20,
      },
      "props": Object {},
      "type": "Other",
    },
    "string example",
  ],
  "props": Object {},
  "type": "Example",
}
`;
